week5/palindromeLL.cc

```diff
@@ -0,0 +1,176 @@
+/* palindromeLL.cc
+ * author: connie
+ *
+ * LINKED LIST PALINDROME
+ * checks if a LL is a palindrome.
+ */
+
+#include <cstdlib>
+#include <iostream>
+#include <cstdio>
+#include <string>
+#include <stack>
+#include <cmath>
+#include <cassert>
+
+using namespace std;
+
+//== DEFINITIONS
+struct node
+{
+  char data;
+  node *next;
+  node(char d) : data(d), next(NULL) {}
+};
+
+//== DECLARATIONS
+void insert(node*& head, int key);
+void deleteLL(node*& head);
+bool palindromeLL(node *head);
+
+//== FUNCTIONS
+/*
+ * void insert(node*& head, int key)
+ * inputs: head - the LL node to insert the new node in front of
+           key - value of the new node to be inserted
+ * output: none
+ * notes: helper function to insert a new node into the LL, as the new head node
+ */
+void insert(node*& head, char d)
+{
+  node *curr = new node(d);
+  curr->next = head;
+
+  head = curr;
+}
+
+/*
+ * void deleteLL(node*& head)
+ * inputs: head - the start of the entire LL that we want to delete
+ * output: none
+ * notes: helper function to prevent memory leaks
+ */
+void deleteLL(node*& head)
+{
+  node *del = head;
+
+  while (del) {
+    head = head->next;
+    delete del;
+    del = head;
+  }
+
+  head = NULL;
+}
+
+
+/*
+ * bool palindromeLL(node *head)
+ * inputs: *head - pointer to the starting node of the LL
+ * output: T if the LL is a palindrome, F if not
+ * notes: main solution function
+ */
+bool palindromeLL(node *head)
+{
+  // case: null or only one -> is a palindrome
+  if (!head || (!head->next)) return true;
+
+  // find the middle node of the LL by using a faster 'runner' node
+  // since runner goes 2x as fast as mid, by the time runner is at the end,
+  // the mid node will be in the middle of the LL
+  node *mid = head;
+  node *runner = head;
+  while (runner && runner->next) {
+    runner = runner->next->next;
+    mid = mid->next;
+  }
+
+  // use a stack for the first half of LL
+  stack<char> palindrome;
+  node *c;
+  for (c = head; c != mid; c = c->next) palindrome.push(c->data);
+  if (runner) c = c->next; // SKIP the middle element if size of LL is odd
+
+  // check remaining half to verify palindrome
+  while (c->next) {
+      if (c->data != palindrome.top()) return false;
+      palindrome.pop();
+      c = c->next;
+  }
+
+  // if we reach this point then we only need to verify the first and last elements
+  return (c->data == palindrome.top());
```

Comment:
Great job with your week 5 hw. You're correct that you can reverse a linked list with O(1) additional space. It's actually not something that we'll go over tomorrow (in fact, you'll see a solution shockingly similar to yours :)), but it's a nice optimization and a great practice problem.